File: task 2/extract.py
# Массивы для русских и латинских букв-замен
rus_letters = ['а', 'е', 'о', 'р', 'с', 'у', 'х', 'А', 'В', 'Е', 'К', 'О', 'Р', 'С', 'Т', 'Х']
eng_letters = ['a', 'e', 'o', 'p', 'c', 'y', 'x', 'A', 'B', 'E', 'K', 'O', 'P', 'C', 'T', 'X']

container_with_hidden_info_file = 'code.txt'
output_secret_message_file = 'message.txt'

def bin_to_text(bin_str):
    """Преобразование бинарного представления в текст с учётом кодировки CP1251."""
    byte_array = bytearray()
    for i in range(0, len(bin_str), 8):
        byte = bin_str[i:i+8]
        byte_array.append(int(byte, 2))
    
    return byte_array.decode('cp1251')

def extract_info_from_container(container):
    """Извлечение скрытой информации из контейнера."""
    secret_message_bin = []
    
    for char in container:
        if char in rus_letters:
            # Если буква осталась русской, значит бит = 0
            secret_message_bin.append('0')
        elif char in eng_letters:
            # Если буква заменена на английский аналог, значит бит = 1
            secret_message_bin.append('1')
    
    return ''.join(secret_message_bin)

# Чтение контейнера
with open(container_with_hidden_info_file, 'r', encoding='cp1251') as container_f:
    container_text = container_f.read()

# Извлечение бинарного сообщения
secret_message_bin = extract_info_from_container(container_text)

# Преобразование бинарного сообщения в текст
secret_message = bin_to_text(secret_message_bin)

# Запись извлеченного сообщения
with open(output_secret_message_file, 'w', encoding='cp1251') as output_f:
    output_f.write(secret_message)

File: task 2/extract.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <bitset>
#include <algorithm>

using namespace std;

const std::vector<wchar_t> rus_letters = {L'а', L'е', L'о', L'р', L'с', L'у', L'х', L'А', L'В', L'Е', L'К', L'О', L'Р', L'С', L'Т', L'Х'};
const std::vector<wchar_t> eng_letters = {L'a', L'e', L'o', L'p', L'c', L'y', L'x', L'A', L'B', L'E', L'K', L'O', L'P', L'C', L'T', L'X'};

std::wstring bin_to_text(const std::wstring& bin_str) {
    std::wstring text;
    for (size_t i = 0; i < bin_str.length(); i += 8) {
        std::bitset<8> byte(bin_str.substr(i, 8));
        text += static_cast<wchar_t>(byte.to_ulong());
    }
    return text;
}

std::wstring extract_info_from_container(const std::wstring& container) {
    std::wstring secret_message_bin;

    for (wchar_t ch : container) {
        if (std::find(rus_letters.begin(), rus_letters.end(), ch) != rus_letters.end()) {
            secret_message_bin += L'0'; // Бит = 0
        } else if (std::find(eng_letters.begin(), eng_letters.end(), ch) != eng_letters.end()) {
            secret_message_bin += L'1'; // Бит = 1
        }
    }

    return secret_message_bin;
}

int main() {
    std::wstring container_with_hidden_info_file = L"code.txt";
    std::wstring output_secret_message_file = L"message.txt";

    std::wifstream container_f(container_with_hidden_info_file);
    std::wofstream output_f(output_secret_message_file);

    if (!container_f.is_open() || !output_f.is_open()) {
        std::wcerr << L"Ошибка при открытии файлов!" << std::endl;
        return 1;
    }

    std::wstring container_text((std::istreambuf_iterator<wchar_t>(container_f)), std::istreambuf_iterator<wchar_t>());
    
    // Извлечение бинарного сообщения
    std::wstring secret_message_bin = extract_info_from_container(container_text);
    
    // Преобразование бинарного сообщения в текст
    std::wstring secret_message = bin_to_text(secret_message_bin);

    output_f << secret_message;

    container_f.close();
    output_f.close();

    return 0;
}
